### PR TITLE
Fix cached decoding in the HF attention adapter

### PR DESCRIPTION
**Pull request by GodRayyy**
## Problem

The HF adapter sends cached inputs to a Triton kernel that assumes equal Q and KV lengths.

With Gemma4-E4B attention shapes (`Hq=8`, `Hkv=2`, `D=256/512`):

- `Q=1, KV=64` fails Triton compilation.
- `Q=16, KV=64` runs but returns wrong outputs. The relative L2 error is about 2.5 compared with SDPA.

## Fix

Use SDPA for unequal Q/KV lengths and sequences shorter than 16 tokens. Align queries with the end of the KV cache. Preserve the sliding window, supplied masks, and attention scale, including Gemma4's `scale=1.0`.

Keep normal equal-length training on Triton. Add a length check to the raw training API to prevent silent wrong results.

## Environment and tests

- NVIDIA RTX 4090 (SM 8.9)
- Python 3.12.12, PyTorch 2.10.0+cu128, Triton 3.6.0
- 11 local tests passed with no skips, including 20 FP16/BF16 GPU cases with output and gradient checks against an FP64 reference.
- GPU coverage includes the original 16 D=256/512 cases and 4 new D=256 cases for truncated sliding-window caches, differently left-padded batches, and single-/multi-token decoding. The new cases compare against a full-history FP64 reference.
- Added regression coverage for 2D padding-mask suffix alignment and rejection of masks shorter than the retained KV length. The deterministic suffix-alignment regression failed before the fix and passes after it.
- Worst GPU gradient relative L2 error in the original 16 GPU cases: 0.00402.

```bash
python -m unittest discover -s tests -p test_cached_attention.py -v
```

Tests cover the attention adapter. Full-model training and generation were not tested. The existing shared-memory limit covered by #4 is outside this fix.

### CODE REVIEW

**Comment by GodRayyy** [comment]
@codex review

File: flash_attn/hf_integration.py [hunk 1]
```diff
@@ -107,6 +108,44 @@ def _compute_image_group_state(mm_token_type_ids: torch.Tensor) -> _ImageGroupSt
...
+    else:
+        mask = None
+        if is_causal:
+            q_pos = torch.arange(nq, device=query.device)[:, None] + nk - nq
+            k_pos = torch.arange(nk, device=query.device)[None, :]
+            mask = k_pos <= q_pos
+            if slide > 0:
+                mask = mask & (k_pos > q_pos - slide)
+        if attention_mask is not None:
+            padding_mask = attention_mask[:, None, None, :nk].bool()
```
[the remark this answers is not in the document]
Comment (GodRayyy) [reply flash_attn/hf_integration.py]: Thanks for catching this. Fixed the 2D padding-mask alignment to use the retained KV suffix, and added a length check to reject masks shorter than the cache. Explicit 4D mask handling is unchanged.

The new regression test failed before the fix. All 11 tests now pass on an RTX 4090, including the original 16 GPU cases and 4 new FP16/BF16 cases checking outputs and gradients against a full-history FP64 reference for single- and multi-token decoding. These tests cover the attention adapter, not full-model generation.

**Comment by GodRayyy** [comment]
@codex review

Addressed the padding-mask alignment issue in 19e1dd4 and added regression tests. All 11 tests pass, including 20 GPU cases.

**Comment by GodRayyy** [comment]
@zzhhjjj This is ready for your review when you have time. The padding-mask issue is fixed in 19e1dd4, all 11 local tests pass on an RTX 4090 (including 20 GPU cases), and Codex re-reviewed the latest commit without finding major issues. Happy to adjust the approach if needed. Thanks!
